loaders/src/glTF/2.0/Extensions/KHR_materials_transmission.ts

```diff
@@ -6,6 +6,258 @@ import { IMaterial } from "../glTFLoaderInterfaces";
+            }
+        }
+        if (this._transparentMeshesCache.length == 0) {
+            this.disabled = true;
+        } else {
+            this.disabled = false;
+        }
+    }
+
+    private _parseScene(): void {
+        this._scene.meshes.forEach(this._addMesh.bind(this));
+        // Listen for when a mesh is added to the scene and add it to our cache lists.
+        this._scene.onNewMeshAddedObservable.add(this._addMesh.bind(this));
+        // Listen for when a mesh is removed from to the scene and remove it from our cache lists.
+        this._scene.onMeshRemovedObservable.add(this._removeMesh.bind(this));
+    }
+
+    // When one of the meshes in the scene has its material changed, make sure that it's in the correct cache list.
+    private onMeshMaterialChanged(mesh: AbstractMesh) {
+        if (mesh instanceof Mesh) {
+            let transparent_idx = this._transparentMeshesCache.indexOf(mesh);
+            let opaque_idx = this._opaqueMeshesCache.indexOf(mesh);
+
+            // If the material is transparent, make sure that it's added to the transparent list and removed from the opaque list
+            const useTransmission = this.shouldRenderAsTransmission(mesh.material);
+            if (useTransmission) {
+                if (mesh.material instanceof PBRMaterial) {
+                    mesh.material.subSurface.refractionTexture = this._opaqueRenderTarget;
+                }
+                if (opaque_idx !== -1) {
+                    this._opaqueMeshesCache.splice(opaque_idx, 1);
+                    this._transparentMeshesCache.push(mesh);
+                } else if (transparent_idx === -1) {
+                    this._transparentMeshesCache.push(mesh);
+                }
+                // If the material is opaque, make sure that it's added to the opaque list and removed from the transparent list
+            } else {
+                if (transparent_idx !== -1) {
+                    this._transparentMeshesCache.splice(transparent_idx, 1);
+                    this._opaqueMeshesCache.push(mesh);
+                } else if (opaque_idx === -1) {
+                    this._opaqueMeshesCache.push(mesh);
+                }
+            }
+        }
+        if (this._transparentMeshesCache.length == 0) {
+            this.disabled = true;
+        } else {
+            this.disabled = false;
+        }
+    }
+
+    /**
+     * Setup the render targets according to the specified options.
+     */
+    private _setupRenderTargets(): void {
+
+        // Remove any layers rendering to the opaque scene.
+        if (this._scene.layers) {
+            this._scene.layers.forEach((layer) => {
+                if (this._opaqueRenderTarget) {
+                    let idx = layer.renderTargetTextures.indexOf(this._opaqueRenderTarget);
+                    if (idx >= 0) {
+                        layer.renderTargetTextures.splice(idx, 1);
+                    }
+                }
+            });
```

Comment:
The if check can be outside of the for loop. Also, consider using for...of loops instead of forEach everywhere since it's a bit cheaper.

Reply:
I think the reason that it's inside the loop is that tslint complains when I put it outside. This is because `this._opaqueRenderTarget` is nullable and `layer.renderTargetTextures.indexOf` doesn't take `null` as a valid type. Is there a nice TS solution to this?

Reply:
Just pass `this._opaqueRenderTarget!` instead of `this._opaqueRenderTarget` if you know that `this._opaqueRenderTarget` can't be null but TS doesn't.

Reply:
If you use for...of, it won't be an issue anymore :)

Reply:
That's one of the down sides to using forEach, not to mention possible perf issues.

Reply:
Thank you for the tips.